crates/paimon/src/spec/types.rs

```diff
@@ -0,0 +1,1039 @@
+    ///
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L84>
+    fn is_any_of(&self, type_roots: &[DataTypeRoot]) -> bool {
+        type_roots.iter().any(|tr: &DataTypeRoot| self.is(tr))
+    }
+
+    /// Returns whether the root of the type is part of at least one family of the families or not.
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L94>
+    fn is_any_of_family(&self, families: &[DataTypeFamily]) -> bool {
+        families
+            .iter()
+            .any(|f: &DataTypeFamily| self.is_with_family(f.clone()))
+    }
+
+    /// Returns a deep copy of this type with possibly different nullability.
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L113>
+    fn copy(&self, is_nullable: bool) -> Self {
+        Self {
+            is_nullable,
+            type_root: self.type_root.clone(),
+        }
+    }
+
+    /// Returns a deep copy of this type. It requires an implementation of {@link #copy(boolean)}.
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L120>
+    fn copy_with_nullable(&self) -> Self {
+        self.copy(self.is_nullable)
+    }
+
+    /// Compare two data types without nullable.
+    /// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/DataType.java#L129>
+    fn copy_ignore_nullable(&self) -> Self {
+        self.copy(false)
+    }
+
+    fn serialize_json(&self) -> String {
+        serde_json::to_string(&self.to_string()).unwrap()
+    }
+
+    fn with_nullability(&self, args: Arguments) -> String {
+        if !self.is_nullable() {
+            format!("{} NOT NULL", args)
+        } else {
+            format!("{}", args)
+        }
+    }
+
+    fn accept<T>(&self, visitor: &mut T)
+    where
+        T: DataTypeVisitor<T>,
+    {
+        visitor.visit(self);
+    }
+
+    fn not_null(&self) -> Self {
+        self.copy(false)
+    }
+
+    fn nullable(&self) -> Self {
+        self.copy(true)
+    }
+}
+
+/// ArrayType for paimon.
+///
+/// Impl Reference: <https://github.com/apache/paimon/blob/db8bcd7fdd9c2705435d2ab1d2341c52d1f67ee5/paimon-common/src/main/java/org/apache/paimon/types/ArrayType.java>.
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, Hash)]
+#[serde(rename_all = "camelCase")]
+pub struct ArrayType {
+    pub element_type: DataType,
+}
+
+impl Display for ArrayType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        if !self.element_type.is_nullable() {
```

Comment:
nit: Maybe use `if self.element_type.is_nullable() { abc } else { def }` for better reading?